File: backend/shared/events.py
"""VESPER Universal Event Schemas & Channel Envelopes.

Defines Pydantic v2 models for client-to-server and inter-service IPC.
All payloads use `extra="allow"` for forward compatibility as modules evolve.
"""

from __future__ import annotations

import time
import uuid
from enum import Enum
from typing import Any, Optional
from pydantic import BaseModel, ConfigDict, Field


# ── Channels ─────────────────────────────────────────────────────────────
class Channel(str, Enum):
    CONTROL = "CONTROL"
    VOICE = "VOICE"
    GESTURE = "GESTURE"
    NOTIFY = "NOTIFY"
    SYSTEM = "SYSTEM"
    AGENT = "AGENT"
    SYNC = "SYNC"


# ── Client Types ─────────────────────────────────────────────────────────
class ClientType(str, Enum):
    DESK_HUD = "DESK_HUD"
    MOBILE_ANDROID = "MOBILE_ANDROID"
    MOBILE_IOS = "MOBILE_IOS"
    EDGE_ORANGE_PI = "EDGE_ORANGE_PI"
    EDGE_NODE = "EDGE_NODE"
    DEV_TEST = "DEV_TEST"
    UNKNOWN = "UNKNOWN"


# ── Standard Event Types ─────────────────────────────────────────────────
class EventType(str, Enum):
    # Control & Handshake
    CLIENT_HELLO = "CLIENT_HELLO"
    SERVER_HELLO = "SERVER_HELLO"
    PING = "PING"
    PONG = "PONG"

    # Voice & Speech
    VOICE_COMMAND = "VOICE_COMMAND"
    VOICE_AUDIO_CHUNK = "VOICE_AUDIO_CHUNK"
    AGENT_RESPONSE = "AGENT_RESPONSE"

    # Gestures
    GESTURE_EVENT = "GESTURE_EVENT"

    # Mobile Notifications
    NOTIFICATION_RELAY = "NOTIFICATION_RELAY"
    NOTIFICATION_DIGEST = "NOTIFICATION_DIGEST"

    # System & Interruption
    SET_VOLUME = "SET_VOLUME"
    ZEN_MODE_STATE = "ZEN_MODE_STATE"
    FOCUS_MODE_STATE = "FOCUS_MODE_STATE"
    MEDIA_CONTROL = "MEDIA_CONTROL"
    INTERRUPT = "INTERRUPT"
    INTERRUPT_ACK = "INTERRUPT_ACK"

    # Cross-Device State Synchronization
    DEVICE_REGISTER = "DEVICE_REGISTER"
    DEVICE_HEARTBEAT = "DEVICE_HEARTBEAT"
    STATE_SYNC = "STATE_SYNC"
    STATE_SNAPSHOT = "STATE_SNAPSHOT"

    # Status / Errors
    ERROR = "ERROR"
    STATUS_UPDATE = "STATUS_UPDATE"



# ── Base Extensible Payload ──────────────────────────────────────────────
class BasePayload(BaseModel):
    """Base model allowing arbitrary extra fields for evolutionary flexibility."""
    model_config = ConfigDict(extra="allow")


# ── Universal Envelopes ──────────────────────────────────────────────────
class ClientEnvelope(BaseModel):
    """Standard message received from a connected client (HUD / Mobile)."""
    uuid: str = Field(default_factory=lambda: str(uuid.uuid4()))
    channel: Channel
    type: EventType | str
    timestamp: float = Field(default_factory=time.time)
    payload: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(extra="allow")


class ServerEnvelope(BaseModel):
    """Standard message emitted by the Gateway to clients."""
    uuid: str = Field(default_factory=lambda: str(uuid.uuid4()))
    channel: Channel
    type: EventType | str
    timestamp: float = Field(default_factory=time.time)
    status: str = "ok"  # "ok" | "error" | "interrupted" | "pending"
    payload: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(extra="allow")


# ── Typed Payload Models (Extensible Helpers) ────────────────────────────

class ClientHelloPayload(BasePayload):
    client_id: str
    client_type: ClientType = ClientType.UNKNOWN
    version: str = "2.0.0"
    capabilities: list[str] = Field(default_factory=list)


class ServerHelloPayload(BasePayload):
    status: str = "authenticated"
    session_id: str
    heartbeat_interval_ms: int = 15000
    server_version: str = "2.0.0"


class PingPayload(BasePayload):
    server_time: float = Field(default_factory=time.time)


class PongPayload(BasePayload):
    client_time: float = Field(default_factory=time.time)


class VoiceCommandPayload(BasePayload):
    command: str
    is_final: bool = True
    confidence: float = 1.0
    image: Optional[str] = None  # Base64 snapshot if vision context is attached


class AudioChunkPayload(BasePayload):
    chunk_index: int = 0
    is_final: bool = False
    format: str = "mp3"  # "mp3" | "pcm" | "wav"
    audio_data: str  # Base64-encoded audio slice


class GestureEventPayload(BasePayload):
    gesture: str  # e.g., "TOGGLE_ZEN", "MUTE", "VOLUME_DIAL:65"
    source: str = "WEBCAM_WORKER"
    confidence: float = 1.0


class NotificationRelayPayload(BasePayload):
    package_name: str
    sender: str
    title: str
    text: str
    is_priority_contact: bool = False


class InterruptPayload(BasePayload):
    target_uuid: Optional[str] = None  # UUID of the active task to cancel (or None for all)
    reason: str = "USER_BARGE_IN"


class SystemVolumePayload(BasePayload):
    volume: int = Field(ge=0, le=100)


class ToggleStatePayload(BasePayload):
    enabled: bool


class ErrorPayload(BasePayload):
    message: str
    code: str = "GENERAL_ERROR"
    details: Optional[dict[str, Any]] = None

File: mobile/src/types/events.ts
/**
 * VESPER Universal Protocol Types for Mobile Companion (Expo).
 * Strictly mirrors backend schemas from `backend/shared/events.py` and `backend/sync/models.py`.
 */

export enum Channel {
  CONTROL = "CONTROL",
  VOICE = "VOICE",
  GESTURE = "GESTURE",
  NOTIFY = "NOTIFY",
  SYSTEM = "SYSTEM",
  AGENT = "AGENT",
  SYNC = "SYNC",
  VISION = "VISION",
}

export enum ClientType {
  DESK_HUD = "DESK_HUD",
  MOBILE_ANDROID = "MOBILE_ANDROID",
  MOBILE_IOS = "MOBILE_IOS",
  EDGE_ORANGE_PI = "EDGE_ORANGE_PI",
  EDGE_NODE = "EDGE_NODE",
  DEV_TEST = "DEV_TEST",
  UNKNOWN = "UNKNOWN",
}

export enum EventType {
  // Control & Handshake
  CLIENT_HELLO = "CLIENT_HELLO",
  SERVER_HELLO = "SERVER_HELLO",
  PING = "PING",
  PONG = "PONG",

  // Voice & Speech Lifecycle
  WAKE_WORD_DETECTED = "WAKE_WORD_DETECTED",
  WAKE_WORD_TOGGLE = "WAKE_WORD_TOGGLE",
  WAKE_WORD_STATE = "WAKE_WORD_STATE",
  VOICE_COMMAND = "VOICE_COMMAND",
  AGENT_ACTIVATING = "AGENT_ACTIVATING",
  AGENT_RESPONSE = "AGENT_RESPONSE",
  AGENT_SPEAKING = "AGENT_SPEAKING",
  AGENT_IDLE = "AGENT_IDLE",

  // Gestures
  GESTURE_EVENT = "GESTURE_EVENT",
  GESTURE_TOGGLE = "GESTURE_TOGGLE",

  // Vision & Remote Display
  SCREEN_CAPTURE_REQUEST = "SCREEN_CAPTURE_REQUEST",
  SCREEN_CAPTURE_RESPONSE = "SCREEN_CAPTURE_RESPONSE",
  CAMERA_FRAME_STREAM = "CAMERA_FRAME_STREAM",

  // Mobile Notifications
  NOTIFICATION_RELAY = "NOTIFICATION_RELAY",
  NOTIFICATION_DIGEST = "NOTIFICATION_DIGEST",

  // System & Interruption
  SET_VOLUME = "SET_VOLUME",
  ZEN_MODE_STATE = "ZEN_MODE_STATE",
  FOCUS_MODE_STATE = "FOCUS_MODE_STATE",
  MEDIA_CONTROL = "MEDIA_CONTROL",
  INTERRUPT = "INTERRUPT",
  INTERRUPT_ACK = "INTERRUPT_ACK",

  // Cross-Device State Synchronization
  DEVICE_REGISTER = "DEVICE_REGISTER",
  DEVICE_HEARTBEAT = "DEVICE_HEARTBEAT",
  STATE_SYNC = "STATE_SYNC",
  STATE_SNAPSHOT = "STATE_SNAPSHOT",

  // Status / Errors
  ERROR = "ERROR",
  STATUS_UPDATE = "STATUS_UPDATE",
}

export interface ClientEnvelope {
  uuid?: string;
  channel: Channel | string;
  type: EventType | string;
  timestamp?: number;
  payload: Record<string, any>;
}

export interface ServerEnvelope {
  uuid: string;
  channel: Channel | string;
  type: EventType | string;
  timestamp: number;
  status: string; // "ok" | "error" | "interrupted" | "pending"
  payload: Record<string, any>;
}

export interface DeviceRegistration {
  device_id: string;
  device_type: string;
  device_name: string;
  hostname?: string;
  os_name?: string;
  architecture?: string;
  is_headless?: boolean;
  has_camera?: boolean;
  has_display?: boolean;
  has_microphone?: boolean;
  battery_level?: number | null;
  is_charging?: boolean | null;
  network_type?: string;
}

export interface SynchronizedState {
  master_volume: number;
  zen_mode: boolean;
  focus_mode: boolean;
  wakeword_active: boolean;
  optical_sensor_active: boolean;
  active_tasks_count: number;
  unread_notifications_count: number;
  recent_notifications: any[];
  current_media: {
    is_playing: boolean;
    track_title: string;
    artist: string;
    progress_ms: number;
  };
  last_speech_summary: string;
  active_devices: Record<string, DeviceRegistration>;
  version: number;
  updated_at: number;
}

export interface MobileNotificationPayload {
  package_name: string;
  app_name?: string;
  title: string;
  text: string;
  subtext?: string;
  priority?: "URGENT" | "HIGH" | "MEDIUM" | "LOW";
  post_time?: number;
}
